File: 0139-word-break/0139-word-break.go
func wordBreak(s string, wordDict []string) bool {
    dp := make([]bool, len(s) + 1)
    dp[len(s)] = true

    for i := len(s) -1; i >= 0; i-- {
        for _, w := range wordDict {
            end := i + len(w)
            if i + len(w) <= len(s) && s[i:end] == w {
                dp[i] = dp[end]
            }
            if dp[i] {
                break
            }
        }
    }

    return dp[0]
}

File: 0139-word-break/0139-word-break.rs
impl Solution {
    pub fn word_break(s: String, word_dict: Vec<String>) -> bool {
        let n = s.len();

        let mut dp = vec![false; n + 1];
        dp[n] = true;

        for i in (0..s.len()).rev() {
            for w in &word_dict {
                let end = i + w.len();

                if end <= n && &s[i..end] == w {
                    dp[i] = dp[end];
                }

                if dp[i] {
                    break;
                }
            }
        }

        dp[0]
    }
}